coordinator/history/history.go

```
@@ -0,0 +1,231 @@
+		store:   newPVStore(&afero.Afero{Fs: afero.NewBasePathFs(osFS, histPath)}),
+		hashFun: sha256.New,
+	}
+	if hashSize != h.hashFun().Size() {
+		return nil, errors.New("mismatch between hashSize and hash function size")
+	}
+	return h, nil
+}
+
+// ConfigureSigningKey sets the signing key for validation and signing of the protected history parts.
+func (h *History) ConfigureSigningKey(signingKey *ecdsa.PrivateKey) {
+	h.signingKey = signingKey
+}
+
+// GetManifest returns the manifest for the given hash.
+func (h *History) GetManifest(hash [hashSize]byte) ([]byte, error) {
+	return h.getContentaddressed("manifests/%s", hash)
+}
+
+// SetManifest sets the manifest and returns its hash.
+func (h *History) SetManifest(manifest []byte) ([hashSize]byte, error) {
+	return h.setContentaddressed("manifests/%s", manifest)
+}
+
+// GetPolicy returns the policy for the given hash.
+func (h *History) GetPolicy(hash [hashSize]byte) ([]byte, error) {
+	return h.getContentaddressed("policies/%s", hash)
+}
+
+// SetPolicy sets the policy and returns its hash.
+func (h *History) SetPolicy(policy []byte) ([hashSize]byte, error) {
+	return h.setContentaddressed("policies/%s", policy)
+}
+
+// GetTransition returns the transition for the given hash.
+func (h *History) GetTransition(hash [hashSize]byte) (*Transition, error) {
+	transitionBytes, err := h.getContentaddressed("transitions/%s", hash)
+	if err != nil {
+		return nil, err
+	}
+	var transition Transition
+	if err := transition.unmarshalBinary(transitionBytes); err != nil {
+		return nil, fmt.Errorf("unmarshaling transition: %w", err)
+	}
+	return &transition, nil
+}
+
+// SetTransition sets the transition and returns its hash.
+func (h *History) SetTransition(transition *Transition) ([hashSize]byte, error) {
+	return h.setContentaddressed("transitions/%s", transition.marshalBinary())
+}
+
+// GetLatest returns the verified transition for the given hash.
+func (h *History) GetLatest() (*LatestTransition, error) {
```

While implementing I realized that I need to be able to tell the difference between generic errors and "no latest set".

---

The issue is that I'd like to update my local state from persistency before handling a SetManifest, when I don't yet have something to CompareAndSet. But I can work around that by ignoring errors when my local state suggests emptiness - I will be stopped at the very latest when my update conflicts. So this would be a good extension, but not necessary right now.

---

use `errors.Is(err, os.ErrNotExist)` ?

---

I can do that for now, but it's somewhat counter to the idea of having pluggable KV stores :)

---

True.